x/supply/keeper/querier.go

```diff
@@ -0,0 +1,76 @@
+package keeper
+
+import (
+	"fmt"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/supply/types"
+)
+
+// NewQuerier creates a querier for supply REST endpoints
+func NewQuerier(k Keeper) sdk.Querier {
+	return func(ctx sdk.Context, path []string, req abci.RequestQuery) (res []byte, err sdk.Error) {
+		switch path[0] {
+		case types.QueryTotalSupply:
+			return queryTotalSupply(ctx, req, k)
+		case types.QuerySupplyOf:
+			return querySupplyOf(ctx, req, k)
+		default:
+			return nil, sdk.ErrUnknownRequest("unknown supply query endpoint")
+		}
+	}
+}
+
+func queryTotalSupply(ctx sdk.Context, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params types.QueryTotalSupplyParams
+
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrInternal(fmt.Sprintf("failed to parse params: %s", err))
+	}
+
+	totalSupply := k.GetSupply(ctx).Total
```

Comment:
Do we really need to paginate supply?

Reply:
well, depends on how many coins does the network have

Reply:
Can you look into generalizing this logic? I've already implemented this twice.

Reply:
agree, should we open an issue for that?

Reply:
Yup, let's do it!